synapse/rest/admin/__init__.py

```diff
@@ -52,25 +52,98 @@
+
+
+    Post request to allow an administrator to add a user.
+    This needs user to have administrator access in Synapse.
+
+    POST /_synapse/admin/v1/users
+    {
+        "username": "user",
+        "password": "secret",
+        "displayname": "User"
+    }
+
+    returns:
+        200 OK with empty object if success otherwise an error.
+    """
 
     def __init__(self, hs):
         self.hs = hs
         self.auth = hs.get_auth()
         self.handlers = hs.get_handlers()
 
     @defer.inlineCallbacks
-    def on_GET(self, request, user_id):
-        target_user = UserID.from_string(user_id)
-        yield assert_requester_is_admin(self.auth, request)
+    def on_GET(self, request):
+        requester = yield self.auth.get_user_by_req(request)
+        yield assert_user_is_admin(self.auth, requester.user)
 
-        if not self.hs.is_mine(target_user):
-            raise SynapseError(400, "Can only users a local user")
+        if not self.hs.is_mine(requester.user):
+            raise SynapseError(400, "Can only list local users")
 
         ret = yield self.handlers.admin_handler.get_users()
-
         return (200, ret)
 
+    @defer.inlineCallbacks
+    def on_POST(self, request):
+        yield assert_requester_is_admin(self.auth, request)
+
+        body = parse_json_object_from_request(request)
+
+        if "username" not in body:
+            raise SynapseError(
+                400, "username must be specified", errcode=Codes.BAD_JSON
+            )
+        else:
+            if (
+                not isinstance(body["username"], text_type)
+                or len(body["username"]) > 512
+            ):
+                raise SynapseError(400, "Invalid username")
+
+        if "password" not in body:
+            raise SynapseError(
+                400, "password must be specified", errcode=Codes.BAD_JSON
+            )
+        else:
+            if (
+                not isinstance(body["password"], text_type)
+                or len(body["password"]) > 512
+            ):
+                raise SynapseError(400, "Invalid password")
+
+        admin = body.get("admin", None)
+        user_type = body.get("user_type", None)
+        displayname = body.get("displayname", body["username"])
+
+        if user_type is not None and user_type not in UserTypes.ALL_USER_TYPES:
+            raise SynapseError(400, "Invalid user type")
+
+        # Reuse the parts of RegisterRestServlet to reduce code duplication
+        from synapse.rest.client.v2_alpha.register import RegisterRestServlet
+
+        register = RegisterRestServlet(self.hs)
+
+        user_id = yield register.registration_handler.register_user(
+            localpart=body["username"].lower(),
+            password=body["password"],
+            admin=bool(admin),
+            default_display_name=displayname,
+            user_type=user_type,
+        )
+
+        result = yield register._create_registration_details(user_id, body)
```

Comment:
this will issue an access token for the user, which I'm not sure is the right thing to do. (even if you do want to end up with an access token for the new user, I think that might be better served via a different API per #6054).

In short, I don't think you need to call _create_registration_details here, which is good since it's meant to be a private method and instantiating a `RegisterRestServlet` here is ugly.